go/src/lib.rs

```diff
@@ -1,9 +1,434 @@
+use glide_core::client::Client as GlideClient;
 /**
  * Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
  */
-use std::ffi::c_void;
+use glide_core::connection_request;
+use protobuf::Message;
+use redis::{cmd, Cmd, FromRedisValue, RedisResult, Value};
+use std::{
+    ffi::{c_void, CStr, CString},
+    os::raw::c_char,
+};
+use tokio::runtime::Builder;
+use tokio::runtime::Runtime;
+
+pub type SuccessCallback =
+    unsafe extern "C" fn(channel_address: usize, message: *const c_char) -> ();
+pub type FailureCallback =
+    unsafe extern "C" fn(channel_address: usize, err_message: *const c_char) -> ();
+
+pub enum Level {
+    Error = 0,
+    Warn = 1,
+    Info = 2,
+    Debug = 3,
+    Trace = 4,
+}
+
+#[repr(C)]
+pub struct ConnectionResponse {
+    conn_ptr: *const c_void,
+    error: *const RedisErrorFFI,
+}
+
+#[repr(C)]
+pub struct RedisErrorFFI {
+    message: *const c_char,
+    error_type: ErrorType,
+}
+
+#[repr(u32)]
+pub enum ErrorType {
+    ClosingError = 0,
+    RequestError = 1,
+    TimeoutError = 2,
+    ExecAbortError = 3,
+    ConnectionError = 4,
+}
+
+pub struct Client {
+    client: GlideClient,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback, // TODO - add specific error codes
+    runtime: Runtime,
+}
+
+fn create_client_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> RedisResult<Client> {
+    let request =
+        connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes).unwrap();
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_name("GLIDE for Redis Go thread")
+        .build()?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime.block_on(GlideClient::new(request)).unwrap(); // TODO - handle errors.
+    Ok(Client {
+        client,
+        success_callback,
+        failure_callback,
+        runtime,
+    })
+}
+
+/// Creates a new client to the given address. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
+#[no_mangle]
+pub extern "C" fn create_client(
+    connection_request: *const u8,
+    request_len: usize,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes = unsafe { std::slice::from_raw_parts(connection_request, request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => {
+            let message_cstring = CString::new(err.to_string()).unwrap();
+            ConnectionResponse {
+                conn_ptr: std::ptr::null(),
+                error: &RedisErrorFFI {
```

Comment:
I see that we are allocating the ConnectionResponse struct and the GlideClient on the heap but I think we also need to allocate RedisErrorFFI and the RedisErrorFFI message on the heap.